contributors/design-proposals/statefulset-update.md

```diff
@@ -0,0 +1,853 @@
+1. All Pods in the StatefulSet are labeled with a 
+`StatefulSetTemplateGenerationLabel` equal to the StatefulSet's 
+`.Status.TargetTemplateGeneration` (This implies they have been created from 
+the PodTemplate at that revision).
+
+When a StatefulSet update is complete, the controller will signal completion by 
+doing the following.
+
+1. The controller will set the StatefulSet's `.Status.TemplateGeneration` to its 
+`.Status.TargetTemplateGeneration`. 
+1. The controller will set the StatefulSet's `Status.CurrentReplicas` to its 
+`Status.UpdatedReplicas`.
+1. The controller will set the StatefulSet's `Status.UpdatedReplicas` to 0.
+
+### Status Reporting
+After processing the creation, update, or deletion of a StatefulSet or Pod, 
+the StatefulSet controller will record its status by persisting the 
+a StatefulSetStatus object. This has two purposes.
+
+1. It allows the StatefulSet controller to recreate the exact StatefulSet 
+membership in the event of a hard restart of the entire system.
+1. It communicates the current state of the StatefulSet to clients. Using the 
+`.Status.ObserverGeneration`, clients can construct a linearizable view of 
+the operations performed by the controller.
+
+When the StatefulSet controller records the status of a StatefulSet it will 
+do the following.
+
+1. The controller will increment the `.Status.ObservedGeneration` to communicate 
+the `.Generation` of the StatefulSet object that was observed.
+1. The controller will set the `.Status.Replicas` to the current number of 
+created Pods.
+1. The controller will set the `.Status.ReadyReplicas` to the current number of 
+Pods that have a Status of Running and a ReadyCondition.
+1. The controller will set the `.Status.TemplateGeneration` and 
+`.Status.TargetTemplateGeneration` 
+in accordance with [maintaining its revision history](#history-maintenance) 
+and the status of any [complete updates](#update-completion).
+1. The controller will set the `.Status.CurrentReplicas` to the number of 
+Pods that it has created from the PodTemplate that corresponds to the 
+current revision of the StatefulSet.
+1. The controller will set the `.Status.UpdatedReplicas` to the number of Pods 
+that it has created from the PodTemplate that corresponds to the target 
+revision of the StatefulSet.
+1. The controller will then persist the StatefulSetStatus make it durable and 
+communicate it to observers.
+
+## API Server
+The API Server will perform validation for StatefulSet updates and ensure that 
+a StatefulSet's `.Spec.TemplateGeneration` is a generator for a strictly 
+monotonically increasing sequence.
+
+### StatefulSet Validation
+As is currently implemented, the API Server will not allow mutation to any 
+fields of the StatefulSet object other than `.Spec.Replicas` and 
+`.Spec.Template.Containers`. This design imposes the following, additional 
+constraints.
+
+1. The `.Spec.RevisionHistoryLimit` must be greater than or equal to `2`. 
```

Comment:
It can be arbitrarily large, but it can't be less than 2. Update can't function without at least 2 PodTemplates in the revision history, the current and the update target.